src/ptw.cc

```diff
@@ -0,0 +1,514 @@
+#include "ooo_cpu.h"
+#include "ptw.h"
+#include "vmem.h"
+
+extern VirtualMemory vmem;
+extern uint64_t current_core_cycle[NUM_CPUS];
+extern uint8_t  warmup_complete[NUM_CPUS];
+extern uint8_t knob_cloudsuite;
+
+template <typename T>
+struct ptw_eq_addr
+{
+    using argument_type = T;
+    const decltype(argument_type::full_v_addr) full_v_addr;
+	const decltype(argument_type::translation_level) translation_level;
+    ptw_eq_addr(decltype(argument_type::address) full_v_addr, decltype(argument_type::translation_level) translation_level) : full_v_addr(full_v_addr), translation_level(translation_level) {}
+    bool operator()(const argument_type &test)
+    {
+        is_valid<argument_type> validtest;
+        if(!validtest(test) || (test.translation_level != translation_level) || (test.returned != INFLIGHT))
+			return false;
+
+		int shift = 12;
+
+		switch(translation_level)
+		{
+			case IS_PTL5: shift+= 9+9+9+9;
+						   break;
+			case IS_PTL4: shift+= 9+9+9;
+						   break;
+			case IS_PTL3: shift+= 9+9;
+						   break;
+	   		case IS_PTL2: shift+= 9;
+	   					   break;
+		}
+
+		return  (((full_v_addr >> shift) & 0x1ff) == ((test.full_v_addr >> shift) & 0x1ff));	
+    }
+};
+
+void PageTableWalker::handle_read()
+{
+	int reads_this_cycle = PTW_MAX_READ;
+
+	bool mshr_full = std::all_of(MSHR.begin(), MSHR.end(), is_valid<PACKET>());
+
+	if(mshr_full)
+		return;
+
+	while(reads_this_cycle > 0)
+	{
+		if((RQ.occupancy() > 0) && (RQ.front().event_cycle <= current_core_cycle[cpu])&& (((CACHE*)lower_level)->RQ.occupancy() < ((CACHE*)lower_level)->RQ.size())) //PTW lower level is L1D
```

Comment:
I think you should rely on the lower level to know whether its RQ is full or not. The lower level will return -2 if the insertion failed. Then, you don't need a cast to `CACHE*`, or even the check here. Also, consider using the function `delay_queue::has_ready()`, since the RQ is a delay queue.